### PR TITLE
Add GHA to publish to maven repository

### LINKED ISSUE #114: Publish snapshots to maven via GHA
Coming from [opensearch-build#3185](https://github.com/opensearch-project/opensearch-build/issues/3185)

### What is happening?

We are de-coupling the task of publishing the maven snapshots from centralized build workflow to individual repositories. What this means is each repo can now publish maven snapshots using GitHub Actions.
This change unblocks the dependent components from waiting for a successful build before they can consume the snapshots. This also ensures that all snapshots are independent and up to date.

### What do you need to do?
1. Determine what artifacts you publish to maven. Most plugins publish only zips. [Example](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/).
2. Add below repository under `publishing` in build.gradle:

```
    repositories {
        maven {
            name = "Snapshots"
            url = "https://aws.oss.sonatype.org/content/repositories/snapshots"
            credentials {
                username "$System.env.SONATYPE_USERNAME"
                password "$System.env.SONATYPE_PASSWORD"
            }
        }
    }
```
[Example](https://github.com/opensearch-project/security/blob/main/build.gradle#L233-L242)

3. Determine what gradle task will publish the content to snapshots repo. Example: `./gradlew publishPluginZipPublicationToSnapshotsRepository`.
4. Create a pull request to set up a GitHub action that will fetch the credentials and publish the snpahots to https://aws.oss.sonatype.org/ [See sample workflow](https://github.com/opensearch-project/security/blob/main/.github/workflows/maven-publish.yml).
5. Add @gaiksaya as a reviewer, who will also take care of creating the IAM role that is required to fetch the credentials and add it to the GitHub Secrets.

Please feel free to reach out to @opensearch-project/engineering-effectiveness.

### PR DESCRIPTION
**Pull request by VijayanB**
### Description
Add GHA to publish to maven repository

### Issues Resolved
#114 

### Check List
- [ ] New functionality includes testing.
    - [x] All tests pass
- [ ] New functionality has been documented.
    - [ ] New functionality has javadoc added
- [ ] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by VijayanB** [comment]
Build failure fix is raised here #129

File: .github/workflows/maven-publish.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        with:
+          role-to-assume: ${{ secrets.PUBLISH_SNAPSHOTS_ROLE }}
+          aws-region: us-east-1
+      - name: publish snapshots to maven
+        run: |
+          export SONATYPE_USERNAME=$(aws secretsmanager get-secret-value --secret-id maven-snapshots-username --query SecretString --output text)
+          export SONATYPE_PASSWORD=$(aws secretsmanager get-secret-value --secret-id maven-snapshots-password --query SecretString --output text)
+          echo "::add-mask::$SONATYPE_USERNAME"
+          echo "::add-mask::$SONATYPE_PASSWORD"
+          ./gradlew publishPluginZipPublicationToSnapshotsRepository
```
Comment (naveentatikonda) [inline .github/workflows/maven-publish.yml:35]: nit: Add new line

**Review by naveentatikonda — APPROVED** [review]

**Review by martin-gaievski — APPROVED** [review]

**Comment by martin-gaievski** [comment]
For reference, there is one more change required in case we need to publish jars in addition to zips, this is PR for such change: https://github.com/opensearch-project/ml-commons/pull/764

File: .github/workflows/maven-publish.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+name: Publish snapshots to maven
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - 'main'
+      - '1.*'
```
Comment (heemin32) [inline .github/workflows/maven-publish.yml:8]: Neural search does not have 1.* branch

**Pull request merged**